File: src/session.py
"""Multi-turn conversation context management for Doubao-TUI.

Wraps the OpenAI-compatible message list in a lightweight :class:`ChatSession`
that supports history mutation, targeted message retrieval, and tool-call
message structures (``tool_calls`` / ``tool`` roles).
"""

from typing import Any, Dict, List

from src.config import Config


class ChatSession:
    """Manages the conversation message history for a single chat session.

    The session always starts with a system message (from
    :attr:`Config.SYSTEM_PROMPT`) and allows appending user / assistant /
    tool-call messages, clearing non-system history, and extracting the latest
    textual assistant reply.

    Attributes:
        _messages: The internal message list in OpenAI chat format.
    """

    def __init__(self) -> None:
        """Initialise a new session with the default system prompt."""
        self._messages: List[Dict[str, Any]] = [
            {"role": "system", "content": Config.SYSTEM_PROMPT}
        ]

    # ---- Public API -----------------------------------------------------------

    def append_message(self, msg: Dict[str, Any]) -> None:
        """Append an arbitrary message dict to the conversation history.

        Use this for ``tool`` results, or assistant messages that contain
        ``tool_calls`` in addition to (or instead of) ``content``.

        Args:
            msg: A message dict following the OpenAI chat format, e.g.
                ``{"role": "tool", "tool_call_id": "...", "content": "..."}``.
        """
        self._messages.append(msg)

    def add_user_message(self, content: str) -> None:
        """Append a user message to the conversation history.

        Args:
            content: The raw text input from the user.
        """
        self._messages.append({"role": "user", "content": content})

    def add_assistant_message(self, content: str) -> None:
        """Append a plain assistant (model) text response.

        Args:
            content: The full response text returned by the model.
        """
        self._messages.append({"role": "assistant", "content": content})

    def get_messages(self) -> List[Dict[str, Any]]:
        """Return the complete message list, including the system prompt.

        Returns:
            A list of message dicts suitable for the OpenAI chat API.
        """
        return list(self._messages)

    def clear_history(self) -> None:
        """Reset the conversation, keeping only the system prompt."""
        self._messages = [
            msg for msg in self._messages if msg.get("role") == "system"
        ]
        # Safety net: re-inject system prompt if it was somehow removed.
        if not self._messages:
            self._messages = [
                {"role": "system", "content": Config.SYSTEM_PROMPT}
            ]

    def get_last_assistant_message(self) -> str:
        """Return the text content of the most recent *textual* assistant reply.

        Assistant messages that only contain ``tool_calls`` (no ``content``)
        are skipped.

        Returns:
            The latest AI text response, or an empty string if no textual
            assistant message exists in the history.
        """
        for msg in reversed(self._messages):
            if msg.get("role") == "assistant" and msg.get("content"):
                return str(msg["content"])
        return ""

File: src/session.js
// ---------------------------------------------------------------------------
// ArkTerm — ChatSession: message history management with sliding window
// ---------------------------------------------------------------------------

class ChatSession {
  /** @param {number} [maxMessages=15] — threshold before compaction triggers */
  constructor(maxMessages = 15) {
    /** @type {Array<{role:string, content:string|null, tool_calls?:Array, tool_call_id?:string}>} */
    this._messages = [];
    this._maxMessages = maxMessages;
    /** Accumulated context summary string (set after first compaction). */
    this._contextSummary = '';
  }

  /** Append a raw message object. */
  appendMessage(msg) {
    this._messages.push(msg);
  }

  /** Add a user message. */
  addUserMessage(text) {
    this._messages.push({ role: 'user', content: text });
  }

  /** Add an assistant text message. */
  addAssistantMessage(text) {
    this._messages.push({ role: 'assistant', content: text });
  }

  /** Return the FULL message list (shallow copy). */
  getMessages() {
    return [...this._messages];
  }

  /**
   * Return a compacted message list suitable for sending to the API.
   *
   * When total messages ≤ _maxMessages, returns all messages unchanged.
   * Otherwise: system prompt + context summary + last N messages.
   *
   * Keeps the most recent REASONING_WINDOW messages (default 6) intact so
   * the model has full context for the immediate conversation thread.
   */
  getCompactMessages(reasoningWindow = 6) {
    const total = this._messages.length;
    if (total <= this._maxMessages) return [...this._messages];

    // ── Identify system message ──────────────────────────────────────────
    const systemIdx = this._messages.findIndex((m) => m.role === 'system');
    const systemMsg = systemIdx >= 0 ? [this._messages[systemIdx]] : [];

    // ── Extract last N messages (the reasoning window) ───────────────────
    const tailStart = Math.max(systemIdx + 1, total - reasoningWindow);
    const tail = this._messages.slice(tailStart);

    // ── Build context summary from truncated middle ──────────────────────
    const middle = this._messages.slice(systemIdx + 1, tailStart);
    const summary = this._buildSummary(middle);

    // If we already have accumulated summary, prepend it
    const combinedSummary = this._contextSummary
      ? this._contextSummary + '\n' + summary
      : summary;
    this._contextSummary = combinedSummary;

    const summaryMsg = {
      role: 'system',
      content: `[Context Summary: Earlier conversation — ${combinedSummary}]`,
    };

    return [...systemMsg, summaryMsg, ...tail];
  }

  /**
   * Build a compact summary from a slice of messages.
   * Extracts file operations, command executions, and key outcomes.
   */
  _buildSummary(messages) {
    const operations = [];
    const filesSeen = new Set();
    const commandsRun = [];

    for (const msg of messages) {
      if (msg.role === 'assistant' && msg.tool_calls) {
        for (const tc of msg.tool_calls) {
          const name = tc.function.name;
          let args = {};
          try { args = JSON.parse(tc.function.arguments || '{}'); } catch { /* */ }

          switch (name) {
            case 'read_file':
              if (args.path && !filesSeen.has('read:' + args.path)) {
                filesSeen.add('read:' + args.path);
              }
              break;
            case 'write_file':
              if (args.path) {
                filesSeen.add('wrote:' + args.path);
                operations.push(`wrote ${args.path}`);
              }
              break;
            case 'patch_file':
              if (args.path) {
                filesSeen.add('patched:' + args.path);
                operations.push(`patched ${args.path}`);
              }
              break;
            case 'execute_command':
              if (args.command) {
                commandsRun.push(args.command.slice(0, 80));
              }
              break;
            case 'view_structure':
              operations.push('listed directory');
              break;
            case 'git_assistant':
              if (args.action === 'commit_all') {
                operations.push('git commit');
              }
              break;
          }
        }
      }
      // Capture tool results for outcomes
      if (msg.role === 'tool' && msg.content) {
        const content = msg.content;
        if (content.startsWith('✅') || content.includes('written') || content.includes('patched') || content.includes('Committed')) {
          // Success — already captured by the assistant message above
        }
        if (content.startsWith('[Error]') || content.startsWith('[Blocked]')) {
          const brief = content.replace(/\n/g, ' ').slice(0, 60);
          operations.push(`ERROR: ${brief}`);
        }
      }
    }

    const parts = [];
    // Unique operations
    const uniqueOps = [...new Set(operations)].slice(0, 8);
    if (uniqueOps.length > 0) {
      parts.push(`Operations: ${uniqueOps.join('; ')}`);
    }
    // Files accessed
    const wroteFiles = [...filesSeen].filter((f) => f.startsWith('wrote:') || f.startsWith('patched:'));
    if (wroteFiles.length > 0) {
      parts.push(`Files modified: ${wroteFiles.map((f) => f.split(':')[1]).join(', ')}`);
    }
    // Commands executed
    const uniqueCmds = [...new Set(commandsRun)].slice(0, 5);
    if (uniqueCmds.length > 0) {
      parts.push(`Commands: ${uniqueCmds.join(' | ')}`);
    }

    return parts.length > 0 ? parts.join('. ') : `${messages.length} messages of conversation`;
  }

  /** Clear all history and reset summary. */
  clearHistory() {
    this._messages = [];
    this._contextSummary = '';
  }

  /** Return the last assistant message content string, or null. */
  getLastAssistantMessage() {
    for (let i = this._messages.length - 1; i >= 0; i--) {
      const m = this._messages[i];
      if (m.role === 'assistant' && typeof m.content === 'string') {
        return m.content;
      }
    }
    return null;
  }
}

module.exports = { ChatSession };
